xla/service/cpu/tests/onednn_convolution_test.cc

```diff
@@ -32,159 +33,216 @@ limitations under the License.
+               ? absl::Substitute(kConvRewriteOptimizationsStr,
+                                  weights_prepacked_, user_scratchpad_)
+               : "";
+  }
 
-TEST_F(ConvolutionTest, Simple3DTestBF16) {
-  if (!IsSupportedType(PrimitiveType::BF16)) {
-    GTEST_SKIP() << "CPU does not support BF16.";
+  std::string ConvStringWithOptimizations(
+      const std::vector<absl::string_view> fused_ops) {
+    std::ostringstream stream;
+    std::for_each(
+        fused_ops.begin(), fused_ops.end(),
+        [&](const absl::string_view& arg) { stream << "\"" << arg << "\","; });
+    std::string fusions = stream.str();
+    if (fused_ops.size() > 0) {
+      fusions.pop_back();
+      return absl::Substitute(kConvRewriteStr,
+                              absl::Substitute(kConvRewriteFusionsStr, fusions),
+                              GetOptimizationsString());
+    }
+    return absl::Substitute(kConvRewriteStr, "", GetOptimizationsString());
   }
 
-  const char* convolution_module_str = R"(
-  HloModule convolution.test.bf16
+  // TODO(intel-tf): Remove this and simplify patterns when Elemental BF16 is
+  // fully supported.
+  PrimitiveType PromotedDtype() {
+    // BF16 is promoted to F32 because not all HLO Instructions currently
+    // support BF16 computations. Meanwhile, FP32 and FP16 elementwise
+    // instructions are not promoted and remain unchanged.
+    return (dtype_ == BF16) ? F32 : dtype_;
+  }
 
-  ENTRY convolution.test.bf16 {
-    p0 = bf16[8,4,5,5,1] parameter(0)
-    p1 = bf16[3,3,3,1,32] parameter(1)
-    ROOT conv = bf16[8,4,5,5,32] convolution(p0, p1), window={size=3x3x3 pad=1_1x1_1x1_1}, dim_labels=b012f_012io->b012f
-})";
+  void AdjustToleranceForDtype(PrimitiveType for_type, float atol, float rtol) {
+    if (dtype_ == for_type) {
+      atol_ = atol;
+      rtol_ = rtol;
+    }
+  }
 
-  EXPECT_TRUE(RunAndCompare(convolution_module_str, ErrorSpec{1e-4, 1e-4}));
-  MatchOptimizedHlo(convolution_module_str, conv_rewrite_str_);
-}
+  std::string PromotedDtypeToString() {
+    return primitive_util::LowercasePrimitiveTypeName(PromotedDtype());
+  }
 
-TEST_F(ConvolutionTest, Simple2DTestF16) {
-  if (!IsSupportedType(PrimitiveType::F16)) {
-    GTEST_SKIP() << "CPU does not support F16.";
+  void RunCompareAndMatchOptimizedHlo(
+      const absl::string_view outline,
+      const std::vector<absl::string_view> fused_ops) {
+    const std::string convolution_module_str =
+        absl::Substitute(outline, dtypeString_, PromotedDtypeToString());
+    EXPECT_TRUE(RunAndCompare(convolution_module_str, ErrorSpec{atol_, rtol_}));
+    MatchOptimizedHlo(convolution_module_str,
+                      ConvStringWithOptimizations(fused_ops));
   }
+};
 
-  const char* convolution_module_str = R"(
-  HloModule convolution.test.f16
+TEST_P(ConvolutionTest, Simple2DTest1) {
+  const absl::string_view outline = R"(
+  HloModule convolution.test
+
+  ENTRY convolution.test {
+    arg.0 = $0[1,22,22,1] parameter(0)
```

Comment:
Same here. I think replacing `$0` with `$dtype` would make it easier to read.

Reply:
Done